File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0)
            return head;
        ListNode* tail = head;
        int len = 1;
        while (tail->next) {
            tail = tail->next;
            len++;
        }
        tail->next = head;
        k = k % len;
        int stepsToNewHead = len - k;

        while (stepsToNewHead--) {
            tail = tail->next;
        }
        head = tail->next;
        tail->next = nullptr;

        return head;
    }
};

File: 0061-rotate-list/0061-rotate-list.py
class Solution:
    def rotateRight(self, head, k):
        if not head or not head.next or k==0:return head
        n=1
        cur=head
        while cur.next:
            cur=cur.next
            n+=1
        cur.next=head
        k%=n
        steps=n-k
        cur=head
        for _ in range(steps-1):
            cur=cur.next
        new_head=cur.next
        cur.next=None
        return new_head